File: smoke_tests/perl/sql_injection.pl
#!/usr/bin/perl
# SQL Injection vulnerabilities in Perl (DBI)

use strict;
use warnings;
use DBI;
use CGI;

my $cgi = CGI->new();
my $dbh = DBI->connect("dbi:mysql:database=test", "user", "pass");

# Test 1: Direct string interpolation in do()
sub vulnerable_do {
    my $id = $cgi->param('id');
    # VULNERABLE: SQL injection via do()
    $dbh->do("DELETE FROM users WHERE id = $id");
}

# Test 2: prepare() with string interpolation
sub vulnerable_prepare {
    my $username = $cgi->param('username');
    # VULNERABLE: SQL injection in prepare
    my $sth = $dbh->prepare("SELECT * FROM users WHERE username = '$username'");
    $sth->execute();
    return $sth->fetchall_arrayref();
}

# Test 3: selectrow_array with interpolation
sub vulnerable_selectrow_array {
    my $email = $cgi->param('email');
    # VULNERABLE: SQL injection
    my @row = $dbh->selectrow_array(
        "SELECT id, name FROM users WHERE email = '$email'"
    );
    return @row;
}

# Test 4: selectrow_hashref
sub vulnerable_selectrow_hashref {
    my $id = $cgi->param('id');
    # VULNERABLE: SQL injection
    my $row = $dbh->selectrow_hashref(
        "SELECT * FROM products WHERE id = $id"
    );
    return $row;
}

# Test 5: selectall_arrayref
sub vulnerable_selectall {
    my $category = $cgi->param('category');
    # VULNERABLE: SQL injection
    my $rows = $dbh->selectall_arrayref(
        "SELECT * FROM products WHERE category = '$category'"
    );
    return $rows;
}

# Test 6: selectcol_arrayref
sub vulnerable_selectcol {
    my $status = $cgi->param('status');
    # VULNERABLE: SQL injection
    my $ids = $dbh->selectcol_arrayref(
        "SELECT id FROM orders WHERE status = '$status'"
    );
    return $ids;
}

# Test 7: LIKE clause injection
sub vulnerable_like {
    my $search = $cgi->param('search');
    # VULNERABLE: SQL injection in LIKE
    my $sth = $dbh->prepare("SELECT * FROM products WHERE name LIKE '%$search%'");
    $sth->execute();
    return $sth->fetchall_arrayref();
}

# Test 8: ORDER BY injection
sub vulnerable_order_by {
    my $sort = $cgi->param('sort');
    # VULNERABLE: SQL injection in ORDER BY
    my $sth = $dbh->prepare("SELECT * FROM products ORDER BY $sort");
    $sth->execute();
    return $sth->fetchall_arrayref();
}

# Test 9: Table name injection
sub vulnerable_table_name {
    my $table = $cgi->param('table');
    # VULNERABLE: SQL injection via table name
    my $sth = $dbh->prepare("SELECT * FROM $table");
    $sth->execute();
    return $sth->fetchall_arrayref();
}

# Test 10: Column name injection
sub vulnerable_column_name {
    my $column = $cgi->param('column');
    # VULNERABLE: SQL injection via column name
    my $sth = $dbh->prepare("SELECT $column FROM users");
    $sth->execute();
    return $sth->fetchall_arrayref();
}

# Test 11: Multiple parameter injection
sub vulnerable_multiple_params {
    my $username = $cgi->param('username');
    my $password = $cgi->param('password');
    # VULNERABLE: Classic login bypass
    my $sth = $dbh->prepare(
        "SELECT * FROM users WHERE username = '$username' AND password = '$password'"
    );
    $sth->execute();
    return $sth->fetchrow_hashref();
}

# Test 12: INSERT injection
sub vulnerable_insert {
    my $name = $cgi->param('name');
    my $email = $cgi->param('email');
    # VULNERABLE: SQL injection in INSERT
    $dbh->do("INSERT INTO users (name, email) VALUES ('$name', '$email')");
}

# Test 13: UPDATE injection
sub vulnerable_update {
    my $id = $cgi->param('id');
    my $status = $cgi->param('status');
    # VULNERABLE: SQL injection in UPDATE
    $dbh->do("UPDATE orders SET status = '$status' WHERE id = $id");
}

# Test 14: Stored procedure injection
sub vulnerable_stored_proc {
    my $param = $cgi->param('param');
    # VULNERABLE: SQL injection in stored procedure call
    $dbh->do("CALL process_order('$param')");
}

# Test 15: execute with interpolated value
sub vulnerable_execute {
    my $id = $cgi->param('id');
    my $sth = $dbh->prepare("SELECT * FROM users WHERE id = ?");
    # Still vulnerable if id is not validated
    $sth->execute($id);
    return $sth->fetchrow_hashref();
}

1;

File: smoke_tests/dart/sql_injection.dart
// SQL Injection vulnerabilities in Dart

import 'package:sqflite/sqflite.dart';

// Test 1: Raw query with string interpolation
Future<void> vulnerableRawQuery(Database db, String userId) async {
  // VULNERABLE: SQL injection via interpolation
  await db.rawQuery('SELECT * FROM users WHERE id = $userId');
}

// Test 2: Raw query with concatenation
Future<void> vulnerableQueryConcat(Database db, String name) async {
  // VULNERABLE: SQL injection via concatenation
  await db.rawQuery('SELECT * FROM users WHERE name = \'' + name + '\'');
}

// Test 3: Raw insert with user data
Future<void> vulnerableRawInsert(Database db, String username, String email) async {
  // VULNERABLE: SQL injection in INSERT
  await db.rawInsert(
    'INSERT INTO users (username, email) VALUES (\'$username\', \'$email\')'
  );
}

// Test 4: Raw update with user data
Future<void> vulnerableRawUpdate(Database db, String status, String id) async {
  // VULNERABLE: SQL injection in UPDATE
  await db.rawUpdate('UPDATE orders SET status = \'$status\' WHERE id = $id');
}

// Test 5: Raw delete with user data
Future<void> vulnerableRawDelete(Database db, String condition) async {
  // VULNERABLE: SQL injection in DELETE
  await db.rawDelete('DELETE FROM logs WHERE $condition');
}

// Test 6: Execute with user input
Future<void> vulnerableExecute(Database db, String tableName) async {
  // VULNERABLE: Table name injection
  await db.execute('DROP TABLE IF EXISTS $tableName');
}

// Test 7: ORDER BY injection
Future<void> vulnerableOrderBy(Database db, String column, String direction) async {
  // VULNERABLE: ORDER BY injection
  await db.rawQuery('SELECT * FROM products ORDER BY $column $direction');
}

// Test 8: LIKE clause injection
Future<void> vulnerableLike(Database db, String search) async {
  // VULNERABLE: LIKE injection
  await db.rawQuery('SELECT * FROM items WHERE name LIKE \'%$search%\'');
}

// Test 9: LIMIT injection
Future<void> vulnerableLimit(Database db, String limit, String offset) async {
  // VULNERABLE: LIMIT/OFFSET injection
  await db.rawQuery('SELECT * FROM data LIMIT $limit OFFSET $offset');
}

// Test 10: UNION injection
Future<void> vulnerableUnion(Database db, String input) async {
  // VULNERABLE: UNION injection
  await db.rawQuery('SELECT id, name FROM users WHERE id = $input');
}

// Test 11: Subquery injection
Future<void> vulnerableSubquery(Database db, String subquery) async {
  // VULNERABLE: Subquery injection
  await db.rawQuery('SELECT * FROM orders WHERE customer_id IN ($subquery)');
}

// Test 12: Column name injection
Future<void> vulnerableColumnName(Database db, String column) async {
  // VULNERABLE: Column name injection
  await db.rawQuery('SELECT $column FROM users');
}

// Test 13: WHERE clause injection
Future<void> vulnerableWhere(Database db, String whereClause) async {
  // VULNERABLE: WHERE clause injection
  await db.rawQuery('SELECT * FROM products WHERE $whereClause');
}

// Test 14: Batch operations with user input
Future<void> vulnerableBatch(Database db, List<String> queries) async {
  // VULNERABLE: Batch of user-controlled queries
  var batch = db.batch();
  for (var query in queries) {
    batch.rawQuery(query);
  }
  await batch.commit();
}

// Test 15: Transaction with user query
Future<void> vulnerableTransaction(Database db, String userId) async {
  // VULNERABLE: SQL injection in transaction
  await db.transaction((txn) async {
    await txn.rawQuery('SELECT * FROM accounts WHERE user_id = $userId');
    await txn.rawUpdate('UPDATE users SET last_login = NOW() WHERE id = $userId');
  });
}
